### PR TITLE
completed week4

### PR DESCRIPTION
**Pull request by haizhou85**

### CODE REVIEW

File: code_challenges.rb [hunk 1]
```diff
@@ -11,9 +11,38 @@
...
+    arr.map do |value|
+        if value.class == Integer && value.abs()%2 == 1
+            finalArr << value
+        end
+    end
+    #sort the array in ascending order
+    return finalArr.sort
+end
+p oddSort fullArr1
+p oddSort fullArr2
```
Comment (rachaelstanislaw) [inline code_challenges.rb:27]: great!

File: code_challenges.rb [hunk 2]
```diff
@@ -23,25 +52,51 @@
...
+        #make sure no negative speed
+        if @speed < y
+            @speed = 0
+        else
+            @speed -= y
+        end
+    end
+end
+p my_bike.pedal_faster 10
+p my_bike.brake 15
```
Comment (rachaelstanislaw) [inline code_challenges.rb:99]: yay!

File: coding-challenges.js [hunk 3]
```diff
@@ -2,59 +2,111 @@
...
+        for(let i = arr.length - 1; i > 0; i--){
+            //find a random index before and includes the element
+            let j = Math.floor(Math.random()*(i+1))
+            //swap the two elements
+            let tmp = arr[i]
+            arr[i] = arr[j]
+            arr[j] = tmp
+        }
+    }
+}
```
Comment (rachaelstanislaw) [inline coding-challenges.js:22]: i like how you go about shuffling!

File: coding-challenges.js [hunk 4]
```diff
@@ -2,59 +2,111 @@
...
+// --------------------2) Create a function that takes in an array of numbers and returns the sum of all the numbers cubed.
+const cubedSum= (nums) =>{
+    //create a varible to store the result
+    let sum = 0
+    //add each cubed number to the sum
+    nums.map(value => {
+        sum += value**3
+    })
+    return sum
+}
```
Comment (rachaelstanislaw) [inline coding-challenges.js:47]: good job!

File: coding-challenges.js [hunk 5]
```diff
@@ -2,59 +2,111 @@
...
-
-
-
+const noDupe = (arr1, arr2) =>{
+    //combine 2 arrays, use set to remove duplicates
+    let setArr = new Set([...arr1, ...arr2])
+    //transfer set to array
+    return [...setArr]
+    
+}
```
Comment (rachaelstanislaw) [inline coding-challenges.js:107]: good use of Set. certainly a very efficient solution. just be prepared for an interviewer to ask you to write out the long way. i always suggest considering an optimized solution but for the sake of practice, i would put a "naive" solution (meaning, less efficient or possibly your initial approach), and an "optimized" solution in your assessment. this practice will have a real-world application as well as it is _extremely_ helpful/impressive in an interview setting. showing prospective employers how you approach a problem and how you go about refactoring is very valuable.

File: interview-questions.md [hunk 6]
```diff
@@ -4,38 +4,38 @@ Answer the following questions. First, without external resources. Challenge you
 
 1. What is an instance variable in Ruby? How is it different from other variables?
 
-  Your answer:
+  Your answer: an instance variable is start with@, it can be only used within the class
```
Comment (rachaelstanislaw) [inline interview-questions.md:7]: yes!

File: interview-questions.md [hunk 7]
```diff
@@ -4,38 +4,38 @@ Answer the following questions. First, without external resources. Challenge you
...
 
-  Researched answer:
+  Researched answer: An instance variable is a variable that is accessible in class, can’t access instance variables from outside the class.
 
 
 
 2. What is a block in Ruby?
 
-  Your answer:
+  Your answer: block is the codes between do and end, it can be wraped within curly brackets.
```
Comment (rachaelstanislaw) [inline interview-questions.md:15]: yes!

File: interview-questions.md [hunk 8]
```diff
@@ -4,38 +4,38 @@ Answer the following questions. First, without external resources. Challenge you
...
 
-  Researched answer:
+  Researched answer:A block is a piece of code that accepts arguments, and returns a value. A block is always passed to a method call.
 
 
 
 3. Ruby has an implicit return. What does that mean?
 
-  Your answer:
+  Your answer: you don't need to write "return" to return the value
```
Comment (rachaelstanislaw) [inline interview-questions.md:23]: yes!

File: interview-questions.md [hunk 9]
```diff
@@ -4,38 +4,38 @@ Answer the following questions. First, without external resources. Challenge you
...
 
 
 
 4. What is object-oriented programming? How is it different than functional programming?
 
-  Your answer:
+  Your answer:Object Oriented Programming is programming with objects. Object are the intersection of data and behavior. 
 
-  Researched answer:
+  Researched answer: Object-oriented programming (OOP) is a way of writing computer programs which is using the idea of "objects" to represent data and methods. Functional Programming is programming with functions. Particularly programming with functions as values. The function is the ultimate tool for abstraction and composition in functional programs. Logic and state is kept separated. You write functions which get data and transform it to other data.
```
Comment (rachaelstanislaw) [inline interview-questions.md:33]: woohoo!

File: interview-questions.md [hunk 10]
```diff
@@ -4,38 +4,38 @@ Answer the following questions. First, without external resources. Challenge you
...
 
-  Researched answer:
+  Researched answer: Object-oriented programming (OOP) is a way of writing computer programs which is using the idea of "objects" to represent data and methods. Functional Programming is programming with functions. Particularly programming with functions as values. The function is the ultimate tool for abstraction and composition in functional programs. Logic and state is kept separated. You write functions which get data and transform it to other data. 
 
 
 
 5. What is the difference between a class and an object?
 
-  Your answer:
+  Your answer: class is the blueprint to create objects.
```
Comment (rachaelstanislaw) [inline interview-questions.md:39]: yay!

**Review by rachaelstanislaw — APPROVED** [review]
GREAT job! keep on commenting your code like you've been doing, it looks excellent! go ahead and merge :)

File: coding-challenges.js (hunk 5 above)
Comment (haizhou85) [reply coding-challenges.js:107]: const noDupe = (arr1, arr2) =>{
    //combine two arrays to a new one
    let wholeArr = [...arr1, ...arr2]
    //create a empty array to receive results
    let finalArr = []
    //loop through the whole array, find those index match the indexOf function,then push into result
    for(let i = 0; i < wholeArr.length; i++){
      if (i == wholeArr.indexOf(wholeArr[i])){
        finalArr.push(wholeArr[i])
      }
    }
    return finalArr
}

I tried another way using indexof to dump the duplicates.

**Comment by rachaelstanislaw** [comment]
Gerat work! Thank you!

**Pull request merged**